### ISSUE TITLE
Jacko: prevent automatic connection to external clients

### ISSUE THREAD

**Issue by A1RO** [csound/csound#1380]
I've written a simple csound mixer using _Jacko_ and _Mixer_ opcodes.
I'm using it with jack2 on linux.

Here's my setup:
![jack_clients](https://user-images.githubusercontent.com/11812395/93036530-1c1f2680-f60e-11ea-80ff-0c920732c123.png)

Everything works great after the initial connections are made, but I'm stuck in a strange dependency loop when I reboot. Jacko seems to expect that _all_ the external jack clients exist when it launches. If _any_ of them doesn't already exist (which is the case for both `mpd` and `mpv`), then csound fails silently. The program still runs, csound jack client gets created along with its ports and connections to existing jack clients , but **no audio can be heard**, from _any_ of the inputs/outputs. To make it work correctly, I need to kill csound process, then launch it again when _every_ external jack clients defined in the *.csd exist.

The thing is, while jacko expects to find the external clients, those clients refuse to start if they don't find jacko (`csound_mixer`).

---

To put it another way, current version of jacko tries to make a connection to an external port for every internal port it creates. From what I understand, we cannot _just_ create internal jack ports (reserved for future connections) without _also_ trying to connect them in one swoop. It's great for hardware devices which keep their jack clients alive as long as they are powered, but not so great for those software which create their jack clients on-the-fly when their own playback starts.

So, to go with _JackoAudioInConnect_ and _JackoAudioOutConnect_ opcodes, I would need something like ___JackoAudioInPortOnly___ and ___JackoAudioOutPortOnly___ which would take a single parameter (internal port name) instead of two.

Am I missing something?

**Comment by hlolli** [comment]
You can connect csound automatically with `-+jack_inportname=[input port name prefix], -+jack_outportname=[output port name prefix]` options, which you can pass as CsOptions xml tag or command line flag. It should accompany `-+rtaudio=jack` with corresponding buffer sizes.

**Comment by vlazzarini** [comment]
yes, the recommended approach is not to use these jacko opcodes, but just use -+rtaudio=jack and output as usual. These opcodes are somewhat non-standard in operation, and I suspect that outside their narrow use case (which I don't quite understand to be fair) they may be buggy.

**Comment by gogins** [comment]
Victor: Do the Jack-related Csound options work with a Jack setup that is
running in "freewheeling" mode?

Regards,
Mike
-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
The rtjack module works as any jack clients with registered callbacks, I don't know if these can be put in freewheeling mode for disk rendering. I suspect not, so that's probably the use case I referred to.
Outside of that I would not recommend using them -- the fact that they interrupt the normal performKmsps() calls is problematic particularly considering most Jack users are trying to be realtime safe.

**Comment by gogins** [comment]
And would the Csound command line options for Jack support connecting
arbitrary Jack input ports to arbitrary Jack output ports? I understand
that a Csound instrument can route audio from any Jack input to any Jack
output, but that's not what I'm asking. I'm asking about configuring a
direct connection, say in the Csound orchestra header.

Regards,
Mike
-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
The rtaudio module has options to connect inputs from named jack sources and outputs to named jack sources, for example. It's not part of the rt module to make any connections that do not have anything to do with Csound. 

I see what you are trying to say here, that there is this or the other feature in these opcodes. However, there's no escaping that part of their implementation is problematic for a general use case. They have not been stress tested beyond a small set of specific uses and I suspect that there may be issues lingering below the surface. So my recommendation stays that if a user wants to use jack for realtime audio IO, the best approach is to use the RT module that Istvan Varga wrote. I do not want to develop this into an argument. It is my opinion and my suggestion, users can take it or leave it.

There remains the possibility of adding features such as connections between external clients etc via options,
and also a freewheeling option. However, we have not had requests for these.

**Comment by gogins** [comment]
How does Csound support multiple Jack client inputs, as in A1RO's example?
It's not clear from the manual.

Regards,
Mike
-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by A1RO** [comment]
Maybe @vlazzarini can confirm, but I would assume [this comment](https://github.com/csound/csound/issues/692#issuecomment-255043156) from 2016 is still relevant.
At least some parts are documented [here](https://csound.com/docs/manual/RealTimeJack.html). Might need some clarifications.

I first tried `-+rtaudio=jack` but I encountered different problems. Tonight I'll try a few different things then report back.

**Comment by vlazzarini** [comment]
if you find issues with +-rtaudio=jack please report on a separate ticket so we can address it there. Best not to get it mixed up.

**Comment by gogins** [comment]
A1RO,

I had the same problems that you are having when I wrote, and used, the
Jacko opcodes.

My solution was to use the system_i opcode in the Csound orchestra header
to kill and restart the Jack daemon and all the external Jack clients, then
to create the Jack configuration in the orchestra header, and then of
course the Csound instruments could use that Jack configuration the same
way every time.

Let me know what happens if you try this.

In your case, if you are not using freewheeling mode and you do not need to
create Jack connections except for the ones in your diagram, i.e. you do
not need Csound to create direct connections between one Jack client and
another, you could try Victor's suggestion.

Regards,
Mike

-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by A1RO** [comment]
@gogins,

In my opinion, a jack server should be up and running for the whole session, i.e. to be always up between computer reboots.
To me (and I suspect many others), it's the backbone of my studio setup.

I'm not using freewheeling mode at the moment, but might be something i'd be interested to try in the future.

To create the connections between my jack clients, I run a simple shell script which is basically a bunch of `jack_connect 'portA' 'portB'`. Its quite reliable. If a connection already exists or is not valid, `jack_connect` just prints a message then goes on. This way, I can call the script on-demand and only the missing connections will be created.

Still, I'd be willing to try both solutions (jack opcodes and `-+rtaudio=jack`) since each has their own strengths and weaknesses.

As of now, I would sum it up like this:
- Jacko allows us to freely connect arbitrary, non-sequential jack ports
- Jacko is more intuitive for jack users, because we only need to match jack server sample rate and buffer size.
- Jacko implementation seems to be non-standard, which might not play well with the rest of csound environment.
- Jacko hangs when I quit the program. I need to kill the process as an extra step when i'm done.

I still believe that `JackoAudioInPortOnly` and `JackoAudioOutPortOnly` are the missing opcodes to complete the feature set.
Do you think it would be a lot of work to implement? I'd be glad to help testing if someone was willing to write some new code for this.

**Comment by gogins** [comment]
If you could email to me a self-contained CSD that would work as you
describe with the JackoAudioInPortOnly and JackoOutPortOnly opcodes you
want, I will try to implement them. I would then also try to debug the hang
at the end.

Regards,
Mike

**Comment by A1RO** [comment]
Looks promising!

Since I kept the file I used for my original post, I wil start from there and add a few comments to explain the intention behind it.
For future reference, I will post it here to help potential new users of csound.

I might be able to do so later today or in the next few days.

**Comment by A1RO** [comment]
Mike, I have a question before going further. There is a very important detail I overlooked.

From [JackoInit manual page](https://csound.com/docs/manual/JackoInit.html):

>  Csound's sr must be equal to the Jack daemon's frames per second.
> Csound's ksmps must be equal to the Jack daemon's frames per period.
> Frames per period must not only (a) be a power of 2, but also (b) go evenly into the frames per second, e.g. 128 frames per period goes into 48000 frames per second 375 times, for a latency or MIDI time granularity of about 2.7 milliseconds (as good as or better than the absolute best human performers).

With my current setup, I cannot go below 256 frames at 48000 frames per second to get a stable system for any significant load.
A latency of 2.7 ms as presented in the example is extremely low in my experience.

With my settings, I suppose samples are lost in the process? (48000 / 256 = 187.5)
I can confirm that it did work with 256 frames, though it might explain the hang...

Since debugging of the `-+rtaudio=jack` goes well, I tend more and more toward this solution for my use case.

**Comment by A1RO** [comment]
Thank you guys for sharing. I'm new to csound and I learned a lot in the last week. It's good to know the community is alive.
